### DISCUSSION TITLE
Credenziali amministratore

### DISCUSSION THREAD

**Discussion by colellagabriele889** [iisgiua/giuaschool#739] [Richieste di aiuto] [answered]
Salve, una volta che vi viene allegato il file dei log, sono visibili i dati d'accesso per l'account amministratore?

**Comment by trinko** [comment, marked as answer]
> Salve, una volta che vi viene allegato il file dei log, sono visibili i dati d'accesso per l'account amministratore?

NO.
I dati visibili riguardano le pagine visitate, l'utente che le visita (ma non la password), l'indirizzo IP, gli errori registrati.
In ogni caso, se trovi nel file di log qualcosa che ritieni non sia opportuno rendere pubblico, lo puoi rimuovere.

**Reply by colellagabriele889** [reply]
Va bene, grazie. Per quanto riguarda alla contribuzione per questo progetto, come si può entrare nel team per contribuire il progetto? Sono disposto ad aiutarvi con questo progetto.

**Reply by trinko** [reply]
> Va bene, grazie. Per quanto riguarda alla contribuzione per questo progetto, come si può entrare nel team per contribuire il progetto? Sono disposto ad aiutarvi con questo progetto.

Hai esperienza di sviluppo con il framework Symfony? Oppure di cosa ti vorresti occupare?

**Reply by colellagabriele889** [reply]
Mi occuperei anche di risolvere i problemi che vengono pubblicati nelle discussioni o nelle segnalazioni di errore.

**Reply by trinko** [reply]
> Mi occuperei anche di risolvere i problemi che vengono pubblicati nelle discussioni o nelle segnalazioni di errore.

Bene, in tal caso, per contribuire è sufficiente che intervieni direttamente nella discussione o nella segnalazione di errore, dando indicazioni per risolvere il problema.
Se vuoi, potresti anche fare un videotutorial per l'installazione su Altervista, cosa che periodicamente viene chiesta dagli utenti.

**Reply by colellagabriele889** [reply]
Benissimo, grazie per l'opportunità. Provvederò appena possibile di eseguire il tutorial.
 Appena puoi, puoi spiegarmi come trovare l'esatto errore nei file log? Anche con un tutorial di un file log qualsiasi. Grazie in anticipo.

**Reply by registro-giuaschool** [reply]
Puoi allegarmi come trovare il problema nei file log tramite e-mail oppure qui. ☺️ @trinko

**Reply by trinko** [reply]
> Puoi allegarmi come trovare il problema nei file log tramite e-mail oppure qui. ☺️ @trinko

Guarda, posso darti solo qualche indicazione, perché la ricerca di un bug dipende fondamentalmente dalla tua conoscenza del PHP e dell'applicazione da controllare.

Comunque, il punto iniziale è sapere quale errore viene generato ed esattamente in quale riga di codice. Questo si ottiene dal log degli errori: solo gli errori _critici_ vengono segnalati all'utente, per cui bisogna cercare nel log le righe che contengono la parola **CRITICAL**.
Nella stessa riga si va poi a cercare l'indicazione dell'errore e della riga di codice che l'ha generato. Esempio:
```
[2024-06-11T22:46:00.998843+02:00] php.CRITICAL: Uncaught Error: array_keys(): Argument #1 ($array) must be of type array, null given {"exception":"[object] (TypeError(code: 0): array_keys(): Argument #1 ($array) must be of type array, null given at /xxx/yyy/src/Util/ScrutinioUtil.php:677)"} {"username":"XXX","roles":["ROLE_STAFF","ROLE_DOCENTE","ROLE_UTENTE"]}
[2024-06-11T22:46:01.000198+02:00] request.CRITICAL: Uncaught PHP Exception TypeError: "array_keys(): Argument #1 ($array) must be of type array, null given" at /xxx/yyy/src/Util/ScrutinioUtil.php line 677 {"exception":"[object] (TypeError(code: 0): array_keys(): Argument #1 ($array) must be of type array, null given at /xxx/yyy/src/Util/ScrutinioUtil.php:677)"} {"client_ip":"X.X.X.X","uri":"http://XXX.altervista.org/public/coordinatore/scrutinio","query_string":null,"method":"GET","username":"XXX","roles":["ROLE_STAFF","ROLE_DOCENTE","ROLE_UTENTE"]}

```

Nell'esempio si vede l'indicazione dell'errore (**array_keys(): Argument #1 ($array) must be of type array, null given**), il file incriminato (**src/Util/ScrutinioUtil.php**) e la riga di codice nel file (**677**). A volte può essere utile sapere anche la pagina che si sta eseguendo (**http://XXX.altervista.org/public/coordinatore/scrutinio**).

A questo punto, sapendo il tipo di errore, si ripercorrono le istruzioni precedenti che possono averlo causato, comprese le chiamate ad altre funzioni presenti anche in altri file.
Come dicevo, la caccia all'errore può essere facilitata dalla propria esperienza di programmazione nel linguaggio usato e nell'applicazione esaminata. Ma non è detto che si arrivi ad una soluzione, anche perché non sappiamo quali sono i dati presenti nel database dell'utente.

Nell'esempio, l'errore è chiaro: si sta usando la funzione _array_keys_ ma viene passato come parametro un valore nullo e non un array. La riga di codice è:
```
    $proposte = $this->em->getRepository('App\Entity\PropostaVoto')->proposte($classe, $periodo,
      array_keys($dati['alunni']), array_keys($dati['materie']));
```
Quindi **$dati['alunni']** o **$dati['materie']** è vuoto anziché contenere i dati che ci si aspetta. 
Andando a vedere come vengono valorizzate queste variabili, ripercorrendo il codice all'indietro, si vede che sono valori letti dal database. Se l'applicazione è configurata correttamente, queste variabili conterranno gli alunni da scrutinare e le materie previste per la classe. 
La cosa più probabile è che l'utente non abbia configurato correttamente il registro, per cui non ci sono alunni associati alla classe o cattedre di docenti per la classe. 
Puoi guardare direttamente come la segnalazione si è conclusa: #721 

**Strumenti da usare.**
E' fondamentale utilizzare un buon editor di programmazione: io consiglio _Visual Studio Code_, che ritengo ottimo ed è usatissimo. 

Naturalmente devi avere sul tuo PC tutta l'applicazione completa in modalità DEV (developer). E l'applicazione deve poter essere avviata in locale (sul tuo pc) in modo da poter riprodurre l'errore e verificare se le eventuali modifiche che effetui lo possono risolvere.

Per questo consiglio di usare i docker, come descritto nella documentazione: [Installazione tramite Docker](https://iisgiua.github.io/giuaschool-docs/install-docker.html)
Unica differenza è che dovrai usare l'immagine per lo sviluppo: ghcr.io/iisgiua/**giuaschool-dev**:latest

**Reply by registro-giuaschool** [reply]
Bene, ho installato correttamente Docker Desktop, adesso cosa devo fare?

**Reply by trinko** [reply]
> Bene, ho installato correttamente Docker Desktop, adesso cosa devo fare?

Per prima cosa prova ad avviare il server. Usando la linea di comando, esegui: 
```
docker run -d --rm --name gs_test -p 443:443 ghcr.io/iisgiua/giuaschool-dev:latest
``` 
Quindi con il browser inserisci l'indirizzo **https://localhost** 
Se è tutto ok, vedi il registro, a cui puoi accedere come amministratore con utente _admin_ e password _admin_.

Noterai che appare una barra di debug in basso, con una serie di icone cliccabili: questo ti permette di vedere molte informazioni che possono essere utili per il debug (ad esempio i comandi SQL eseguiti sul database).

**Reply by registro-giuaschool** [reply]
Ecco cosa risulta appena inserisco il comando nella barra di comando. 
![17214020184946674912280194640924](https://github.com/user-attachments/assets/adfb2354-c23d-41bf-9596-d4b0e74c25d4)
![Uploading 172140206524384223800555438092.jpg…]()

**Reply by trinko** [reply]
Tu usi Windows, quindi devi caricare il componente WSL2 (che è una macchina virtuale con Linux).

Il problema mi sembra essere che la virtualizzazione non è abilitata. Devi entrare nel BIOS del tuo PC e attivarla.
